Kernel/Devices/Device.cpp

```diff
@@ -15,21 +17,69 @@
 
 namespace Kernel {
 
-Device::Device(MajorNumber major, MinorNumber minor)
-    : m_major(major)
-    , m_minor(minor)
+struct AllDevicesDetails {
+    SpinlockProtected<HashMap<u64, BlockDevice*>, LockRank::None> block_devices {};
+    SpinlockProtected<HashMap<u64, CharacterDevice*>, LockRank::None> char_devices {};
+    SpinlockProtected<CircularQueue<DeviceEvent, 100>, LockRank::None> event_queue {};
+    // NOTE: There's no locking on this pointer because we expect to initialize it once
+    // and never touch it again.
+    OwnPtr<BaseDevices> base_devices;
+};
+
+static Singleton<AllDevicesDetails> s_all_details;
+
+SpinlockProtected<CircularQueue<DeviceEvent, 100>, LockRank::None>& Device::event_queue()
+{
+    return s_all_details->event_queue;
+}
+
+BaseDevices* Device::base_devices()
+{
+    return s_all_details->base_devices.ptr();
+}
+
+UNMAP_AFTER_INIT void Device::initialize_base_devices()
 {
+    auto base_devices = MUST(try_make<BaseDevices>(*NullDevice::must_initialize(), *ConsoleDevice::must_create(), *DeviceControlDevice::must_create()));
+    s_all_details->base_devices = move(base_devices);
+}
+
+RefPtr<Device> Device::acquire_by_type_and_major_minor_numbers(DeviceNodeType type, MajorNumber major, MinorNumber minor)
+{
+    VERIFY(type == DeviceNodeType::Block || type == DeviceNodeType::Character);
+
+    if (type == DeviceNodeType::Block) {
+        return s_all_details->block_devices.with([&](auto& map) -> RefPtr<Device> {
+            auto it = map.find(encoded_device(major.value(), minor.value()));
+            if (it == map.end())
+                return nullptr;
+            return *it->value;
+        });
+    }
+
+    return s_all_details->char_devices.with([&](auto& map) -> RefPtr<Device> {
+        auto it = map.find(encoded_device(major.value(), minor.value()));
+        if (it == map.end())
+            return nullptr;
+        return *it->value;
+    });
```

Comment:
I know this is preexisting, but it feels kind of sad to duplicate the whole thing just for the different pointer type.

Reply:
We can probably write an helper lambda here, right?
What do you think about this?

Reply:
It would have to be a generic one I guess, given that the inner type is also different. But yeah, if that works that would be great.